### ISSUE TITLE
[CRASH] keydb-server v6.3.2 eat memory and stuck

### ISSUE THREAD

**Issue by rnz** [Snapchat/KeyDB#568]
**Crash report**

![image](https://user-images.githubusercontent.com/851927/218777469-609256b4-3543-49c8-975d-69363d87af01.png)

```
490973:555999:S 14 Feb 2023 01:42:44.059 * RDB: 6958 MB of memory used by copy-on-write
490973:555999:S 14 Feb 2023 01:42:44.729 # tombstones removed: 1111077, remain: 0
490973:491079:S 14 Feb 2023 01:42:47.269 * Background saving terminated with success
490973:491079:S 14 Feb 2023 01:42:47.370 * Starting automatic rewriting of AOF on 133% growth
490973:491079:S 14 Feb 2023 01:42:47.634 * Background append only file rewriting started by pid 556088
490973:491079:S 14 Feb 2023 01:43:39.328 * AOF rewrite child asks to stop sending diffs.
556088:491079:C 14 Feb 2023 01:43:39.328 * Parent agreed to stop sending diffs. Finalizing AOF...
556088:491079:C 14 Feb 2023 01:43:39.328 * Concatenating 1247.47 MB of AOF diff received from parent.
490973:491079:S 14 Feb 2023 01:43:42.723 * Background AOF buffer size: 80 MB
556088:491079:C 14 Feb 2023 01:43:43.071 * SYNC append only file rewrite performed
556088:491079:C 14 Feb 2023 01:43:43.277 * AOF rewrite: 3695 MB of memory used by copy-on-write
490973:491079:S 14 Feb 2023 01:43:43.814 * Background AOF rewrite terminated with success
490973:491079:S 14 Feb 2023 01:43:43.843 * Residual parent diff successfully flushed to the rewritten AOF (109.14 MB)
490973:491079:S 14 Feb 2023 01:43:43.843 * Background AOF rewrite finished successfully
490973:491079:S 14 Feb 2023 01:44:20.343 * 10000000 changes in 60 seconds. Saving...
490973:491079:S 14 Feb 2023 01:44:20.346 * Background saving started
490973:556252:S 14 Feb 2023 01:45:12.180 * DB saved on disk
490973:556252:S 14 Feb 2023 01:45:12.358 * RDB: 6920 MB of memory used by copy-on-write
490973:556252:S 14 Feb 2023 01:45:12.637 # tombstones removed: 1009501, remain: 0
490973:491079:S 14 Feb 2023 01:45:14.833 * Background saving terminated with success
490973:491079:S 14 Feb 2023 01:46:30.447 * Starting automatic rewriting of AOF on 100% growth
490973:491079:S 14 Feb 2023 01:46:30.697 * Background append only file rewriting started by pid 556435
490973:491079:S 14 Feb 2023 01:46:35.102 * Background AOF buffer size: 80 MB
490973:491079:S 14 Feb 2023 01:46:38.654 * Background AOF buffer size: 180 MB
490973:491079:S 14 Feb 2023 01:46:41.976 * Background AOF buffer size: 280 MB
490973:491079:S 14 Feb 2023 01:46:46.071 * Background AOF buffer size: 380 MB
490973:491079:S 14 Feb 2023 01:46:50.193 * Background AOF buffer size: 480 MB
...
490973:491079:S 14 Feb 2023 02:47:59.561 # Background AOF buffer size: 90980 MB
490973:491079:S 14 Feb 2023 02:48:03.560 * Background AOF buffer size: 91080 MB
490973:491079:S 14 Feb 2023 02:48:07.563 * Background AOF buffer size: 91180 MB
490973:491079:S 14 Feb 2023 02:48:11.538 * Background AOF buffer size: 91280 MB
490973:491079:S 14 Feb 2023 02:48:15.534 * Background AOF buffer size: 91380 MB
556435:signal-handler (1676342900) Received SIGTERM scheduling shutdown...
556435:signal-handler (1676342905) Received SIGTERM scheduling shutdown...
```
SIGTERM signals sended by OOMKiller

**Additional information**

1. Distributor ID:	Debian
Description:	Debian GNU/Linux 11 (bullseye)
Release:	11
Codename:	bullseye

2. Steps to reproduce (I don't think it can be reproduced, but here are the conditions under which KeyDB crash):
Schema:
```
standalone1 (master, 17mln keys, keydb 6.0.18) 
      |
      |
standalone2  (replica, keydb 6.3.2 - it was stuck)
      |
      |
clusternode1 (multi-master, replica of standalone2)
      |
      ...
      |
clusternode5 (multi-master)
```

Config standalone2:
```
root@kdb10-prod-xprj:~# grep -E '^#|^$' -v /etc/keydb/keydb.conf
bind * -::*
protected-mode yes
port 8379
tcp-backlog 2048
timeout 0
tcp-keepalive 300
daemonize yes
pidfile /var/run/keydb/keydb-server.pid
loglevel notice
logfile /var/log/keydb/keydb-server.log
databases 128
always-show-logo yes
set-proc-title yes
proc-title-template "{title} {listen-addr} {server-mode}"
save 900 100000
save 300 500000
save 60 10000000
stop-writes-on-bgsave-error yes
rdbcompression yes
rdbchecksum yes
dbfilename dump.rdb
rdb-del-sync-files no
dir /var/lib/keydb
replica-serve-stale-data yes
replica-read-only no
repl-diskless-sync no
repl-diskless-sync-delay 5
repl-diskless-load disabled
repl-timeout 600
repl-disable-tcp-nodelay no
replica-priority 100
acllog-max-len 128
lazyfree-lazy-eviction yes
lazyfree-lazy-expire yes
lazyfree-lazy-server-del yes
replica-lazy-flush yes
lazyfree-lazy-user-del yes
lazyfree-lazy-user-flush yes
oom-score-adj no
oom-score-adj-values 0 200 800
disable-thp yes
appendonly yes
appendfilename "appendonly.aof"
appendfsync everysec
no-appendfsync-on-rewrite no
auto-aof-rewrite-percentage 100
auto-aof-rewrite-min-size 64mb
aof-load-truncated yes
aof-use-rdb-preamble yes
lua-time-limit 5000
slowlog-log-slower-than 10000
slowlog-max-len 128
latency-monitor-threshold 0
notify-keyspace-events ""
hash-max-ziplist-entries 512
hash-max-ziplist-value 64
list-max-ziplist-size -2
list-compress-depth 0
set-max-intset-entries 512
zset-max-ziplist-entries 128
zset-max-ziplist-value 64
hll-sparse-max-bytes 3000
stream-node-max-bytes 4096
stream-node-max-entries 100
activerehashing yes
client-output-buffer-limit normal 0 0 0
client-output-buffer-limit replica 0 0 0
client-output-buffer-limit pubsub 32mb 8mb 60
hz 10
dynamic-hz yes
aof-rewrite-incremental-fsync yes
rdb-save-incremental-fsync yes
jemalloc-bg-thread yes
server-threads 14
active-replica no
multi-master no
replicaof rdb05-prod-xprj 6379
replica-weighting-factor 2
```

**Comment by msotheeswaran** [comment]
This is due to AOF child hanging and buffer growing forever, fixed with #560.

**Comment by rnz** [comment]
I think this error is more interesting - when keydb receives a SIGTERM from OOMkiller, KeyDB shutdown all threads, but the main process remains alive, and this is wrong behavior, because OOMkiller does not try to kill KeyDB's main process, and SystemD does not receive an exitcode and does not trying to start KeyDB again
